### PR TITLE
Added die operator

### LINKED ISSUE #28: Declaring errors in templates
Some templates may require that an error be thrown if a certain condition is met. This could be a missing data member or perhaps another condition. Currently there is no way to throw an exception that "escapes" the Horseshoe parser because all exceptions are caught by `Expression.java:915`.

My suggestion is to add a sort of "exception literal" such as the currently unused apostrophe `'`. This literal would function like parenthesis, but it would capture the value of an expression as an exception. For example `'"This is an exception: " + myVariable.toString()'`. These "exception literals" would construct a new Java exception class `TemplateException` that is thrown by the `Template.render` method. "Exception literals" are only thrown if they are the return value of an expression, therefore the expression `{{'"This is an unthrown exception"'; "Hello world!"}}` would not throw an `TemplateException`.

More specifically the syntax would be: `'` `Expression` `'`.

### LINKED ISSUE #31: Recursive named expressions
I created a simple recursive named expression test that calculates the Fibonacci sequence for a given input number, `{{fib -> . > 1 ? fib(. - 1) + fib(. - 2) : .}}`. `{{fib(0)}}` and `{{fib(1)}}` both return the correct values, all other results are empty. See 5741f9d5f.

### PR DESCRIPTION
**Pull request by nicklauslittle**
Closes #28.
Closes #31.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,7 +57,8 @@ Horseshoe does not support Mustache lambdas. It foregoes lambdas in favor of an
 | 13 | a<code>&#124;&#124;</code>b \(Logical Or\) | Left&nbsp;to&nbsp;right |
 | 14 | a<code>?:</code>b \(Null Coalesce\), <br>a<code>??</code>b \(Null Coalesce \(Alternate\)\), <br>a<code>?</code>b \(Ternary\), <br>a<code>:</code>b \(Pair\) | Right&nbsp;to&nbsp;left |
 | 15 | a<code>=</code>b \(Assign\), <br>a<code>\*=</code>b \(Multiply Assign\), <br>a<code>/=</code>b \(Divide Assign\), <br>a<code>%=</code>b \(Modulus Assign\), <br>a<code>\+=</code>b \(Add Assign\), <br>a<code>\-=</code>b \(Subtract Assign\), <br>a<code>&amp;=</code>b \(Bitwise And Assign\), <br>a<code>^=</code>b \(Bitwise Xor Assign\), <br>a<code>&#124;=</code>b \(Bitwise Or Assign\), <br>a<code>&lt;&lt;=</code>b \(Bitwise Shift Left Assign\), <br>a<code>&gt;&gt;=</code>b \(Bitwise Shift Right Sign Extend Assign\), <br>a<code>&gt;&gt;&gt;=</code>b \(Bitwise Shift Right Zero Extend Assign\), <br>a<code>?:=</code>b \(Null Coalesce Assign\), <br>a<code>??=</code>b \(Null Coalesce Assign \(Alternate\)\) | Right&nbsp;to&nbsp;left |
-| 16 | a<code>,</code>b\* \(Array / Map Separator\), <br>a<code>;</code>b \(Statement Separator\) | Left&nbsp;to&nbsp;right |
+| 16 | <code>☠</code>a \(Die\) | Left&nbsp;to&nbsp;right |
```
Comment (tbateson) [inline README.md]: How is anyone supposed to type this? It isn't the same as the emoji skull and crossbones (☠️).

File: src/main/java/horseshoe/Template.java [hunk 2]
```diff
@@ -78,13 +82,18 @@ Section getSection() {
...
+			renderContext.getIndentation().push("");
+			renderContext.getSectionData().push(renderContext.getGlobalData());
 
-		for (final Action action : actions) {
-			action.perform(renderContext, writer);
+			for (final Action action : actions) {
+				action.perform(renderContext, writer);
+			}
+		}
+		catch (final HaltRenderingException e) {
```
Comment (tbateson) [inline src/main/java/horseshoe/Template.java]: same lines as previous brace.

File: src/main/java/horseshoe/internal/HaltRenderingException.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package horseshoe.internal;
+
+public class HaltRenderingException extends RuntimeException {
```
Comment (tbateson) [inline src/main/java/horseshoe/internal/HaltRenderingException.java]: Not sure this needs to be public, package private should be acceptable.

File: README.md (hunk 1 above)
Comment (tbateson) [reply README.md]: Disregard the second half of that comment, VSCode was acting bizarrely. This does seem odd to use an emoji as an operator, does any other language do something like this?

Comment (nicklauslittle) [reply README.md]: No, I just ran out of operators.

File: src/main/java/horseshoe/internal/HaltRenderingException.java (hunk 3 above)
Comment (nicklauslittle) [reply src/main/java/horseshoe/internal/HaltRenderingException.java]: Must be public, used in horseshoe.internal.Expression and horseshoe.Template.

**Comment by tbateson** [comment]
I've pushed another commit that has a multi-function version of the Fibonacci sequence. I'm not sure if your intention is to support what I'm doing in the template, but I would think that forward declaration of named expressions in not required.

**Comment by nicklauslittle** [comment]
Merged.

**Pull request closed without merging**